File: julia/2022/day_08.jl
function lines_of_sight(grid::Matrix, index::CartesianIndex)::Vector
    row, col = Tuple(index)
    [
        grid[row-1:-1:1, col],
        grid[row+1:end, col],
        grid[row, col-1:-1:1],
        grid[row, col+1:end],
    ]
end

function is_visible(grid::Matrix, index::CartesianIndex)::Bool
    any(all(line .< grid[index]) for line in lines_of_sight(grid, index))
end

function line_score(tree_height::UInt, line::Vector{UInt})::UInt
    score = 0
    for height in line
        score += 1
        if height >= tree_height
            break
        end
    end
    score
end

function scenic_score(grid::Matrix, index::CartesianIndex)::UInt
    val = grid[index]
    prod(line_score(val, line) for line in lines_of_sight(grid, index))
end

function parse_line(line::AbstractString)::Vector{UInt}
    [parse(UInt, c) for c in line]
end

function main()
    text = read("input/2022/input_08.txt", String)
    grid = split(text) .|> parse_line |> Base.splat(hcat)
    num_visible = count(is_visible(grid, i) for i in CartesianIndices(grid))
    println("Part 1: $num_visible")
    high_score = maximum(scenic_score(grid, i) for i in CartesianIndices(grid))
    println("Part 2: $high_score")
end

main()

File: kotlin/y2022/Day08.kt
package y2022

import java.io.File

fun linesOfSight(grid: List<List<Int>>, position: Pair<Int, Int>): Array<List<Int>> {
    val (y, x) = position
    return arrayOf(
        grid[y].slice(0..<x).reversed(),
        grid[y].slice(x+1..<grid[y].size),
        (0..<y).reversed().map { grid[it][x] },
        (y+1..<grid.size).map { grid[it][x] },
    )
}

fun isVisible(grid: List<List<Int>>, position: Pair<Int, Int>): Boolean {
    val height = grid[position.first][position.second]
    return linesOfSight(grid, position).any { line -> line.all { it < height } }
}

fun lineScore(treeHeight: Int, line: Iterable<Int>): Int {
    var score = 0
    for (height in line) {
        score += 1
        if (height >= treeHeight) break
    }
    return score
}

fun scenicScore(grid: List<List<Int>>, position: Pair<Int, Int>): Int {
    val height = grid[position.first][position.second]
    return linesOfSight(grid, position).map { lineScore(height, it) }.reduce(Int::times)
}

fun main() {
    val text = File("input/2022/input_08.txt").readText().trim()
    val grid = text.lines().map { it.map(Char::digitToInt) }
    val positions = grid.withIndex().flatMap { (y, row) -> row.indices.map { x -> Pair(y, x) } }
    val numVisible = positions.count { isVisible(grid, it) }
    println("Part 1: $numVisible")
    val highScore = positions.maxOf { scenicScore(grid, it) }
    println("Part 2: $highScore")
}
